lib/reversetunnel/api_with_roles.go

```diff
@@ -0,0 +1,110 @@
+/*
+Copyright 2020 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package reversetunnel
+
+import (
+	"context"
+
+	"github.com/gravitational/teleport/lib/auth"
+	"github.com/gravitational/teleport/lib/services"
+
+	"github.com/gravitational/trace"
+)
+
+// NewServerWithRoles returns new authorizing server
+func NewServerWithRoles(server Server, roles services.RoleSet, ap auth.AccessPoint) *ServerWithRoles {
+	return &ServerWithRoles{
+		server: server,
+		roles:  roles,
+		ap:     ap,
+	}
+}
+
+// ServerWithRoles authorizes requests
+type ServerWithRoles struct {
+	server Server
+	// roles is a set of roles used to check RBAC permissions.
+	roles services.RoleSet
+	// ap is access point
+	ap auth.AccessPoint
+}
+
+// GetSites returns a list of connected remote sites
+func (s *ServerWithRoles) GetSites() ([]RemoteSite, error) {
+	clusters, err := s.server.GetSites()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	out := make([]RemoteSite, 0, len(clusters))
+	for _, cluster := range clusters {
+		if _, ok := cluster.(*localSite); ok {
+			out = append(out, cluster)
+			continue
+		}
+		rc, err := s.ap.GetRemoteCluster(cluster.GetName())
```

Comment:
this is O(N) requests to the auth server
can you fetch all remote clusters and filter them against local `RemoteSite`s?
or will this always be served from the cache?

Reply:
yes, this is always served from cache that is using in memory `BTree`